internal/controller/standalone_pgadmin/statefulset.go

```diff
@@ -33,6 +35,27 @@ func (r *PGAdminReconciler) reconcilePGAdminStatefulSet(
 	ctx context.Context, pgadmin *v1beta1.PGAdmin,
 	configmap *corev1.ConfigMap, dataVolume *corev1.PersistentVolumeClaim,
 ) error {
+	// In a previous version of PGO, the PGAdmin StatefulSet was not assigned a Parallel Pod Management Policy
+	// which meant that in certain scenarios, a failed Pod could not recover (e.g. a bad image) and would have
+	// to be manually deleted. We now assign the Parallel Pod Management Policy, however, this is not a property
+	// that can be patched, so if we run into a StatefulSet from the previous PGO version that does not have the
+	// parallel policy in place, we must delete and recreate the StatefulSet.
+	existing := &appsv1.StatefulSet{ObjectMeta: naming.StandalonePGAdmin(pgadmin)}
+	if err := errors.WithStack(r.Client.Get(ctx, client.ObjectKeyFromObject(existing), existing)); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		if existing.Spec.PodManagementPolicy != appsv1.ParallelPodManagement {
```

Comment:
🤔 The other function is the one that chooses the parallel policy. Perhaps it would be tidier to generate the STS and compare the existing object to that one?

```go
	sts := statefulset(r, pgadmin, configmap, dataVolume)

	// Previous versions of PGO used a StatefulSet update strategy and policy that
	// could leave the pod in a failed state. When we see it has the wrong policy,
	// delete the StatefulSet while leaving its pods in place. They will be
	// claimed by the StatefulSet created during the next reconciliation.
	// - https://docs.k8s.io/concepts/workloads/controllers/statefulset/#forced-rollback
	var existing appsv1.StatefulSet
	err := r.Client.Get(ctx, client.ObjectKeyFromObject(sts), &existing)

	if err == nil && existing.Spec.PodManagementPolicy != sts.Spec.PodManagementPolicy {
		// delete, 🤔 always return early?
	}

	if err := errors.WithStack(r.setControllerReference(pgadmin, sts)); err != nil {
		return err
	}
	return errors.WithStack(r.apply(ctx, sts))
```

Reply:
Do we not care about the error if the return from `r.Client.Get` isn't `nil`? Obviously if it is a "NotFound" error then we don't care, but if it's a different error?

I didn't make it return early when deleting because I figured that we wanted to do `r.apply(ctx, sts)` regardless... Is it a bad idea to do both during one reconciliation cycle?

Reply:
> Do we not care about the error if the return from `r.Client.Get` isn't `nil`? Obviously if it is a "NotFound" error then we don't care, but if it's a different error?

I couldn't decide. I think I was being too clever. We don't want to have to think about it, so returning it is best.

> I didn't make it return early when deleting because I figured that we wanted to do `r.apply(ctx, sts)` regardless... Is it a bad idea to do both during one reconciliation cycle?

It might not be gone yet. 🤔 We don't have a policy of trying to "do just one thing," but maybe that would be good. I trust that controller-runtime will wake us back up when the DELETE event comes through.